### PR TITLE
add hoprec and ask for code review

### PR DESCRIPTION
**Pull request by ALLEN-CHUANG**

### CODE REVIEW

File: hub/hoprec.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    unsigned long long total_update_times = (unsigned long long)update_times*1000000;
+    unsigned long long worker_update_times = total_update_times/worker;
+    unsigned long long finished_update_times = 0;
+    Monitor monitor(total_update_times);
+
+    // main
+    // 0. [FileGraph] read graph
+    std::cout << "(UI-Graph)" << std::endl;
+	FileGraph file_graph(train_path, 0);
```
Comment (chihming) [inline hub/hoprec.cpp]: 4 blanks for the indent, but not tab

File: hub/hoprec.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    #pragma omp parallel for
+    for (int w=0; w<worker; w++)
+    {
+        long user, item_pos, item_neg;
+        std::vector<double> user_loss(dimension, 0.0);
+        std::vector<double> item_loss(dimension, 0.0);
+        std::vector<long> neighbor;
+        unsigned long long update=0, report_period = 10000;
+        double alpha=init_alpha, alpha_min=alpha*0.0001;
+        user_loss.resize(dimension, 0.0);
```
Comment (chihming) [inline hub/hoprec.cpp]: redundant

File: hub/hoprec.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        item_loss.resize(dimension, 0.0);
```
Comment (chihming) [inline hub/hoprec.cpp]: redundant

File: hub/hoprec.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        while (update < worker_update_times)
+        {
+            int up_flag=0.0;
```
Comment (chihming) [inline hub/hoprec.cpp]: rename the variable with readable meaning

File: hub/hoprec.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  	ArgParser arg_parser(argc, argv);
+    std::string train_path = arg_parser.get_str("-train", "", "train graph path");
+    std::string save_name = arg_parser.get_str("-save", "bpr.embed", "path for saving mapper");
+    int dimension = arg_parser.get_int("-dimension", 64, "embedding dimension");
+    int num_negative = arg_parser.get_int("-num_negative", 5, "number of negative sample");
+    double update_times = arg_parser.get_double("-update_times", 10, "update times (*million)");
+    double init_alpha = arg_parser.get_double("-init_alpha", 0.01, "init learning rate");
+    double l2_reg = arg_parser.get_double("-l2_reg", 0.025, "l2 regularization");
+    double margin = arg_parser.get_double("-margin", 1.0, "update margin");
+    int walk_steps = arg_parser.get_int("-walk_steps", 2, "number of walking steps");
```
Comment (chihming) [inline hub/hoprec.cpp]: rename this `num_hub`?

File: hub/hoprec.cpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            // 4.1 sample positive (user, item_pos) pair, feed the loss, update
+            user = sampler.draw_a_vertex();
+            neighbor = neighbor_sampler.draw_a_walk(user, walk_steps*2);
+
+            // 4.2 sampler negative (user, item_pos, item_neg) pair, feed the loss, update
+            for (int walk=0; walk<walk_steps; walk++)
+            {
+                if(walk==0){
+                    for (int n=0; n<num_negative; n++){
+                        item_pos = sampler.draw_a_context(user);
```
Comment (chihming) [inline hub/hoprec.cpp]: `item_pos` shall be the same as `neighbor[0]`?

File: hub/hoprec.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        double alpha=init_alpha, alpha_min=alpha*0.0001;
+        user_loss.resize(dimension, 0.0);
+        item_loss.resize(dimension, 0.0);
+
+        while (update < worker_update_times)
+        {
+            int up_flag=0.0;
+            // 4.1 sample positive (user, item_pos) pair, feed the loss, update
+            user = sampler.draw_a_vertex();
+            neighbor = neighbor_sampler.draw_a_walk(user, walk_steps*2);
```
Comment (chihming) [inline hub/hoprec.cpp]: rename this variable

File: hub/hoprec.cpp (hunk 6 above)
Comment (ALLEN-CHUANG) [reply hub/hoprec.cpp]: yes, so I simplify the code.

**Pull request merged**